pkg/pillar/cmd/downloader/syncop.go

```diff
@@ -322,28 +322,54 @@ func sourceFailureError(ip, ifname, url string, err error) {
 	zedcloud.ZedCloudFailure(ifname, url, 1024, 0, false)
 }
 
-func getDatastoreCredential(dst types.DatastoreConfig) (zconfig.CredentialBlock, error) {
-	cred := zconfig.CredentialBlock{}
-	if !dst.IsCipher {
-		cred.Identity = dst.ApiKey
-		cred.Password = dst.Password
-		return cred, nil
+func getDatastoreCredential(ctx *downloaderContext, dst types.DatastoreConfig) (zconfig.CredentialBlock, error) {
+	cipherBlock := new(types.CipherBlockStatus)
+	*cipherBlock = dst.CipherBlockStatus
+	log.Infof("%s, cipherblock decryption, using ciper-context: %s\n",
+		dst.Key(), cipherBlock.CipherContextID)
+	if !cipherBlock.IsCipher {
+		return handleCipherBlockError(ctx, cipherBlock, dst.ApiKey,
+			dst.Password, nil)
 	}
-	if !dst.IsValidCipher {
-		errStr := fmt.Sprintf("%s, Cipher Block is not ready", dst.Key())
+	if len(cipherBlock.Error) != 0 {
+		errStr := fmt.Sprintf("%s, cipherblock is not ready, %s",
+			dst.Key(), cipherBlock.Error)
 		log.Errorln(errStr)
-		return cred, errors.New(errStr)
+		err := errors.New(errStr)
+		return handleCipherBlockError(ctx, cipherBlock, dst.ApiKey,
+			dst.Password, err)
 	}
-	clearData, err := tpmmgr.DecryptCipherBlock(dst.CipherBlock)
+	clearData, err := tpmmgr.DecryptCipherBlock(dst.CipherBlockStatus)
 	if err != nil {
-		log.Errorf("%s, dataStore CipherData decryption failed, %v\n",
-			dst.Key(), err)
-		return cred, err
+		log.Errorf("%s, cipherblock decryption failed, %v\n", dst.Key(), err)
+		return handleCipherBlockError(ctx, cipherBlock, dst.ApiKey,
+			dst.Password, err)
 	}
+	cred := zconfig.CredentialBlock{}
 	if err := proto.Unmarshal(clearData, &cred); err != nil {
-		log.Errorf("%s, dataStore Credential unmarshall failed, %v\n",
-			dst.Key(), err)
-		return cred, err
+		log.Errorf("%s, credential unmarshall failed, %v\n", dst.Key(), err)
+		return handleCipherBlockError(ctx, cipherBlock, dst.ApiKey,
+			dst.Password, err)
 	}
+	ctx.pubCipherBlockStatus.Publish(cipherBlock.Key(), cipherBlock)
+	log.Infof("%s, cipherblock decryption successful\n", dst.Key())
 	return cred, nil
 }
+
+func handleCipherBlockError(ctx *downloaderContext,
```

Comment:
I would be useful to have a comment about this function. I only know what it does because of what Saurabh told me today.

Reply:
Done.